contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
 
 	PG_RETURN_NULL();
 }
+
+int 
+SPI_execute_raw_parsetree(RawStmt *parsetree, bool read_only, long tcount)
+{
+	_SPI_plan			plan;
+	int					ret;
+	List				*plancache_list;
+	CachedPlanSource	*plansource;
+	int					prev_sql_dialect;
+
+	if (parsetree == NULL || tcount < 0)
+		return SPI_ERROR_ARGUMENT;
+	
+	/*
+	 * set sql_dialect to tsql, which is needed for raw parsetree parsing 
+	 * and processing
+	 */
+	prev_sql_dialect = sql_dialect;
+	sql_dialect = SQL_DIALECT_TSQL;
+	
+	memset(&plan, 0, sizeof(_SPI_plan));
+	plan.magic = _SPI_PLAN_MAGIC;
+	plan.parse_mode = RAW_PARSE_DEFAULT;
+	plan.cursor_options = CURSOR_OPT_PARALLEL_OK;
+
+	/*
+	 * Construct plancache entries, but don't do parse analysis yet.
+	 */
+	plancache_list = NIL;
+
+	/* 
+	 * src sql can be optained from pstate->p_sourcetext, but
+	 * it is not important here
+	 */
+	plansource = CreateOneShotCachedPlan(parsetree,
+										"SQL NOT AVAILABLE",
+										CreateCommandTag(parsetree->stmt));
+
+	plancache_list = lappend(plancache_list, plansource);
+	plan.plancache_list = plancache_list;
+	plan.oneshot = true;
+
+	ret = SPI_execute_plan_with_paramlist(&plan, NULL, read_only, tcount);
+
+	/* reset sql_dialect */
+	sql_dialect = prev_sql_dialect;
+
+	return ret;
+}
+
+PG_FUNCTION_INFO_V1(bbf_pivot);
+Datum
+bbf_pivot(PG_FUNCTION_ARGS)
+{	
+	ReturnSetInfo   *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc		tupdesc;
+	MemoryContext 	per_query_ctx;
+	MemoryContext 	oldcontext;
+	HTAB	   	   	*bbf_pivot_hash;
+
+	MemoryContext 	tsql_outmost_context;
+	PLtsql_execstate 	*tsql_outmost_estat;
+	RawStmt	   		*bbf_pivot_src_sql;
+	RawStmt	   		*bbf_pivot_cat_sql;
+	int				nestlevel;
+	List			*per_pivot_list;
+	
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context that cannot accept a set")));
+	if (!(rsinfo->allowedModes & SFRM_Materialize) ||
+		rsinfo->expectedDesc == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not allowed in this context")));
+
+	tsql_outmost_estat = get_outermost_tsql_estate(&nestlevel);
+	tsql_outmost_context = tsql_outmost_estat->stmt_mcontext_parent;
+	if (!tsql_outmost_context)
+		ereport(ERROR,
+			(errcode(ERRCODE_SYNTAX_ERROR),
+				errmsg("pivot outer context not found")));
+
+	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
+	per_pivot_list = list_nth_node(List, tsql_outmost_estat->pivot_parsetree_list, tsql_outmost_estat->pivot_number - 1);
+	bbf_pivot_src_sql = list_nth_node(RawStmt, per_pivot_list, 0);
+	bbf_pivot_cat_sql = list_nth_node(RawStmt, per_pivot_list, 1);
+	MemoryContextSwitchTo(oldcontext);
```

Comment:
Add a comment what's happening here. 
Also, add Assert check on the length of the lists you're trying to access.

For instance:
```
Assert(list_length(per_pivot_list) >= 2); 
bbf_pivot_src_sql = list_nth_node(RawStmt, per_pivot_list, 0);
bbf_pivot_cat_sql = list_nth_node(RawStmt, per_pivot_list, 1);
```